src/lib/provable/gadgets/twisted-curve.ts

```diff
@@ -0,0 +1,451 @@
+  },
+  toBigint({ x, y }: Point) {
+    let x_ = Field3.toBigint(x);
+    let y_ = Field3.toBigint(y);
+    return { x: x_, y: y_, infinity: x_ === 0n && y_ === 1n };
+  },
+  isConstant: (P: Point) => Provable.isConstant(Point, P),
+
+  /**
+   * Random point on the curve.
+   */
+  random(Curve: CurveTwisted) {
+    return Point.from(random(Curve));
+  },
+
+  provable: provable({ x: Field3, y: Field3 }),
+};
+
+function add(
+  p1: Point,
+  p2: Point,
+  Curve: { modulus: bigint; a: bigint; d: bigint }
+) {
+  let { x: x1, y: y1 } = p1;
+  let { x: x2, y: y2 } = p2;
+  let f = Curve.modulus;
+  let a = Curve.a;
+  let d = Curve.d;
+
+  // constant case
+  if (Point.isConstant(p1) && Point.isConstant(p2)) {
+    let p3 = twistedAdd(Point.toBigint(p1), Point.toBigint(p2), f, a, d);
+    return Point.from(p3);
+  }
+
+  assert(
+    Curve.modulus > l2Mask + 1n,
+    `Base field moduli smaller than ${l2Mask + 1n} are not supported`
+  );
+
+  // witness and range-check denominators, x3, y3
+  let witnesses = exists(12, () => {
+    let [x1_, x2_, y1_, y2_] = Field3.toBigints(x1, x2, y1, y2);
+
+    // TODO: reuse code in twistedAdd to avoid recomputing these
+
+    let x1x2 = mod(x1_ * x2_, f);
+    let y1y2 = mod(y1_ * y2_, f);
+    let x1y2 = mod(x1_ * y2_, f);
+    let y1x2 = mod(y1_ * x2_, f);
+    let ax1x2 = mod(a * x1x2, f);
+
+    let x3Num = mod(x1y2 + y1x2, f);
+    let y3Num = mod(y1y2 - ax1x2, f);
+
+    let dx1x2y1y2 = mod(d * x1x2 * y1y2, f);
+
+    let x3Denom = inverse(mod(1n + dx1x2y1y2, f), f) ?? 0n;
+    let y3Denom = inverse(mod(1n - dx1x2y1y2, f), f) ?? 0n;
+
+    let x3 = mod(x3Num * x3Denom, f);
+    let y3 = mod(y3Num * y3Denom, f);
+
+    return [...split(x3Denom), ...split(y3Denom), ...split(x3), ...split(y3)];
+  });
+  let [dx0, dx1, dx2, dy0, dy1, dy2, x30, x31, x32, y30, y31, y32] = witnesses;
+  let x3Den: Field3 = [dx0, dx1, dx2];
+  let y3Den: Field3 = [dy0, dy1, dy2];
+  let x3: Field3 = [x30, x31, x32];
+  let y3: Field3 = [y30, y31, y32];
+  ForeignField.assertAlmostReduced([x3Den, x3, y3], f);
+  ForeignField.assertAlmostReduced([y3Den], f);
+
+  // the formula for point addition is well defined for curves in use,
+  // so we don't need to check that the denominators are non-zero
+
+  // x3 = (x1 * y2 + y1 * x2) / (1 + d * x1 * x2 * y1 * y2)
+  // y3 = (y1 * y2 - a * x1 * x2) / (1 - d * x1 * x2 * y1 * y2)
+
+  let one = Field3.from(1n);
+  let x1y2 = ForeignField.mul(x1, y2, f);
+  let y1x2 = ForeignField.mul(y1, x2, f);
+  let x3Num = ForeignField.add(x1y2, y1x2, f);
+  let y1y2 = ForeignField.mul(y1, y2, f);
+  let x1x2 = ForeignField.mul(x1, x2, f);
+  let ax1x2 = ForeignField.mul(Field3.from(a), x1x2, f);
+  let y3Num = ForeignField.sub(y1y2, ax1x2, f);
+
+  let x1x2y1y2 = ForeignField.mul(x1x2, y1y2, f);
+  let dx1x2y1y2 = ForeignField.mul(Field3.from(d), x1x2y1y2, f);
+  // check denominators are correctly computed:
+  // den = 1 / (1 +- d * x1^2 * y1^2)
+  Provable.equal(
```

Comment:
* in general, you should use `Provable.assertEqual()` instead of `Provable.equal(...).assertTrue()`. the latter unnecessarily uses constraints to compute the boolean, while assertEqual is free (just wiring)
* in this case though, why not use `ForeignField.assertMul()`? that should save the constraints for range-checking the addition result